File: app/models/node.rb
class Node < ActiveRecord::Base
  belongs_to :social_network

  has_many :actor_roles, class_name: "Role", foreign_key: "actor_id"
  has_many :relation_roles, class_name: "Role", foreign_key: "relation_id"
  has_and_belongs_to_many :families
  has_many :node_attributes

  attr_accessible :name, :x, :y, :social_network_id, :kind, :family_ids, :actor_role_ids, :relation_role_ids, :node_attribute_ids

  def roles
    if is_actor?
      actor_roles
    else
      relation_roles
    end
  end

  def roles=(new_roles)
    if is_actor?
      actor_roles = new_roles
    else
      relation_roles = new_roles
    end
  end

  def is_actor?
    kind == "Actor"
  end

  def is_relation?
    kind == "Relation"
  end

  def uri(base = social_network.prefix)
    @uri ||= social_network.uri(base) + "/nodes/#{id}"
  end
end

File: app/assets/javascripts/models/node.js
App.Node = DS.Model.extend({
  // attributes
  name: DS.attr('string'),
  kind: DS.attr('string'),
  x: DS.attr('number'),
  y: DS.attr('number'),

  // relationships
  social_network: DS.belongsTo('App.SocialNetwork'),
  families: DS.hasMany('App.Family'),
  actor_roles: DS.hasMany('App.Role', { inverse: 'actor' }),
  relation_roles: DS.hasMany('App.Role', { inverse: 'relation' }),
  node_attributes: DS.hasMany('App.NodeAttribute'),

  // computed attributes
  radius: function(){
    return 12;
  }.property(),
  cx: function () {
    return this.get('x');
  }.property('x'),
  cy: function () {
    return this.get('y') + this.get('radius');
  }.property('y'),
  text_x: function () {
    return this.get('x');
  }.property('x'),
  text_y: function () {
    return this.get('y') + 3*this.get('radius');
  }.property('y'),
  isActor: function() {
    return this.get('kind') == "Actor"; 
  }.property(),

  roles: function() {
    if (this.get('kind') === "Actor") {
      return this.get('actor_roles');
    } else {
      return this.get('relation_roles');
    }
  }.property('actor_roles', 'relation_roles'),
});
